Add API test harness for Linux & OSX

ghubstan opened the pull request:
This PR adds a new `:apitest` subproject to Bisq with the following goals:

* Enable test driven development of the API project
* Automate end to end testing of the API 
* Simplify `:desktop` testing in regtest/full-dao mode

It requires the installation of bitcoin-core v0.19 or v0.20 binaries in the $PATH, or the path to them defined in the subproject's config file `src/main/resources/apitest.properties`:

    bitcoinPath=/home/you/bitcoin/v0.19.1/src
	

The `:apitest` build downloads and installs dao-setup files for `bitcoind`, bob & alice datadirs in `apitest/build/resources/main`, and test case `@BeforeAll` methods start `bitcoind`, `seednode`, `arbitration` node (`:desktop` or `:daemon`), plus bob & alice nodes (`:desktop` or `:daemon`) as Linux background processes.  Currently, the only testing mode supported is regtest / full-dao mode, without Tor.

To run a full clean build, and install dao setup files:

    ./gradlew clean build :apitest:installDaoSetup

To install or re-install dao setup files:

    ./gradlew :apitest:installDaoSetup

To run API `@Test` cases:

    ./gradlew :apitest:test -DrunApiTests=true

A gradle Test Summary containing any initialization errors and test failures can be found in

    apitest/build/reports/tests/test/index.html

This test harness works on a 2014 Mac mini with a Dual-Core i5 and 8 Gb RAM.

`[EDITED  27-July-2020]`

ghubstan committed:
Add configuration support

ApiTestConfig works like :common Config, but specific to this subproject.

BisqAppConfig is an enumeration specifying Bisq desktop and daemon
options for running seednode, arbnode, bob & alices nodes in regtest /
full-dao mode.

ghubstan committed:
Add main resource files

* apitest.properties - config file for customizing ApiTestConfig
  options

* logback.xml - logging config file, will override logback files
  found in classpath

* bitcoin.conf - bitcoin-core regtest config file, overwritten
  during startup with correct path to blocknofity

* blocknotify - bitcoin-core blocknotify config file

ghubstan committed:
Add script to get Bisq app pid

Finding the pid of background linux/java processes has been
difficult to do from java, so a bash script is provided to
simplify the task.

ghubstan committed:
Support starting bitcoin & bisq apps on Linux

The apitest.linux package is for running random bash commands,
running 'bitcoind -regtest', running random 'bitcoin-cli -regtest'
commands, and spinning up Bisq apps such as seednode, arbnode,
and bob & alice nodes.

All but random bash and bitcoin-cli commands are run in the background.

The bitcoin-cli response processing is crude;  a more sophiticated
bitcoin-core rpc interface is not in the scope of this PR.

ghubstan committed:
Add :apitest main driver, setup task & dummy tests

The driver class uses an ExecutorService to submit Callable
tasks for starting bitcoind and Bisq nodes as Linux background
processes.

By default, ApiTestMain starts background processes to support
regtest/dao testing, runs a few bitcoin-cli commands, then
shuts down all background processes before exiting.
(Actual API test suites have not been implemented yet.)

ApiTestConfig options can be used to skip tests and/or leave
background processes running indefinitely.

ghubstan committed:
Add build tasks for installing dao-setup files

This gradle file is 'applied' by the main build file.

Usage:

  Run a full clean, build, download dao-setup.zip,
  and install the zip files contents in directory
  apitest/build/resources/main:

    ./gradlew clean build :apitest:installDaoSetup

  Download (if necessary) the dao-setup.zip file
  and install its contents in directory
  apitest/build/resources/main  (no build).

    ./gradlew :apitest:installDaoSetup

ghubstan committed:
Add subproject :apitest to gradle build file

ghubstan committed:
Fix codacy problems

Unnecessary use of fully qualified name 'System.exit' due to existing
static import 'java.lang.System.exit'.  (line 100)

Avoid throwing raw exception types.  (lines 295, 302)

ghubstan committed:
Fix codacy problems

Avoid throwing raw exception types.

Combine nested if statements.

ghubstan committed:
Fix codacy problem

Avoid throwing null pointer exceptions.

wiz commented:
Does this require adding new jar deps?

ghubstan commented:
> Does this require adding new jar deps?

`[EDITED 19-July-2020]`

The `org.junit.jupiter` dependendency was upgraded to v5.6.2 -- only in the `:apitest:test` dependency definitions, nowhere else in the bisq main or test dependency definitions.  The upgrade is necessary for running ordered tests.

dmos62 commented on build.gradle:
```diff
@@ -143,6 +144,16 @@ configure([project(':cli'),
+                // files in the classpath (:daemon & :cli).
+                def script = file("${rootProject.projectDir}/bisq-$applicationName")
+                script.text = script.text.replace(
+                    'DEFAULT_JVM_OPTS=""', 'DEFAULT_JVM_OPTS="-XX:MaxRAM=6g' +
```
6GB MaxRAM gives me pause.

- Would `ErgoHeapSizeLimit` [0] be a more direct way to limit RAM usage? `MaxRAM` is meant to specify real RAM size (all available memory).
- In the comment it says that we're increasing RAM allotment, so we're doing this because otherwise MaxRAM would be set to something lower?
- What if we're running this on a memory restricted environment (< 6GB RAM)?
- Is this to do with bisq using too much RAM in some cases? But wasn't that only relevant for long-running processes?

[0] http://hg.openjdk.java.net/jdk9/jdk9/hotspot/file/b756e7a2ec33/src/share/vm/runtime/globals.hpp#l2034

ghubstan committed:
Fix error msg spacing

ghubstan committed:
Fix error msgs

dmos62 commented on apitest/src/main/java/bisq/apitest/linux/BitcoinDaemon.java:
```diff
@@ -0,0 +1,91 @@
+        String bitcoindCmd = "export LD_LIBRARY_PATH=" + config.berkeleyDbLibPath + ";"
+                + " " + config.bitcoinPath + "/bitcoind"
+                + " -datadir=" + config.bitcoinDatadir
+                + " -daemon";
```
`export LD_LIBRARY_PATH=` doesn't seem to be necessary with my distribution of `bitcoind`. Also, `config.bitcoinPath` seems to expect to be pointed to the `bin` sub-directory of bitcoin-core's home (the one that contains the `bitcoind` executable). That's a bit unexpected. I think it's more intuitive for it to point to the executable directly, or alternatively to bitcoin-core's home directory.

ghubstan replied:
You no longer need to configure a `BerkeleyDb` lib path if your bitcoin-core binary is statically linked to `BerkeleyDb`.  The `berkeleyDbLibPath` option now defaults to an empty string.  If not set, no `export LD_LIBRARY_PATH` will not be pre-pended to the `bitcoind` command.

ghubstan committed:
Remove MaxRAM from DEFAULT_JVM_OPTS

There is no need for this as any JavaFX based :desktop instances
will be started as background linux processes.

ghubstan committed:
Assume bitcoin-core is statically linked to berkeley-db

The `berkeleyDbLibPath` option now defaults to an empty string.
If not set to a berkeley lib path on the command line or the
apitest.properties file, this option is ignored, and 'bitcoind'
will be started without first exporting the berkeley db library
path.

In other words:  If the bitcoind binary is dynamically linked to
berkeley db libs, export the configured berkeley-db lib path before
starting 'bitcoind'.  If statically linked,  the berkeley db lib
path will not be exported.

Also fixed exception msgs to show missing config file's absolute path.

ghubstan committed:
Fix hanging background process problem

SetupTask submissions for Bisq background apps seednode, arbnode,
etc., would not always complete due to a blocking stderr stream
handler thread.join() call.  This change makes waiting on a bash
process err stream optional.

ghubstan committed:
Add line break in front of port config

ghubstan committed:
Move test setup scaffolding into new Scaffold class

ApiTestMain will run all defined tests, but we also want to run
individual test suites and test cases, and they will need to
run the setup tasks as well.

ghubstan committed:
Expose grpc service stubs

Added :proto to the :apitest classpath for access to grpc
service stubs (to be) used in method (unit) tests.  Added new
GrpcStubs class to expose the grpc service stubs to method and
scenario tests.

The larger goal of :apitest is end to end testing, where :cli's
console output is checked for correctness.

This change partially addresses two other important use cases:

 * "method" testing -- an analog to unit testing
 * "scenario" testing -- an analog to narrow functional testing

For example, tests in the apitest.method package will directly
call grpc services, and asserts will be made on the return values
instead of console output.

Tests in the apitest.scenario package will check correctness
for broader use cases, such as funding a wallet, encrypting then
unlocking a wallet  for a specific time frame, or checking error
messages from the server when a "getbalance" call is made after
an "unlockwallet" timeout has expired.

The broader end to end tests will not use grpc stubs.

ghubstan committed:
Fix codacy problems

Avoid throwing raw exception types.

Document empty method body.

ghubstan committed:
Fix codacy problem

Codacy wants comments inside an empty method.

ghubstan committed:
Support @Skip on test classes and methods

Add super class for all test types (method, scenario, end-to-end),
and an class & method level annotation for skipping tests.

ghubstan committed:
Add first method test cases

These are method test cases for gRPC methods that have already been
well tested by :cli/test.sh

ghubstan committed:
Add MethodTestSuite

ghubstan committed:
Update bats version test

ghubstan committed:
Make init() method public

ghubstan committed:
Add license note, format tearDown(), fix comment

ghubstan committed:
Run MethodTestSuite

ghubstan committed:
Add license note

ghubstan committed:
Fix codacy problem

Add comment inside empty method.

dmos62 commented on apitest/src/main/java/bisq/apitest/config/BisqAppConfig.java:
```diff
@@ -0,0 +1,123 @@
+            BisqAppMain.class.getName(),
+            8888,
+            5123,
+            -1);
```
How do these MaxRAM settings relate to our discussion about similar settings in build.gradle?

ghubstan replied:
Limiting RAM for `bisq-apitest` was not necessary -- and removed from the `bisq-apitest` script -- because it is a headless app.  But we need to limit RAM for `:desktop` apps started by `bisq-apitest`, like we do for the Linux Bisq UI.

Above, you mentioned:
>> Is this to do with bisq using too much RAM in some cases? But wasn't that only relevant for long-running processes?

The Bisq UI is using too much without this limit, especially when starting it on an empty datadir.  This is why `BisqAppConfig` has a larger limit for `bobdesktop` than `bobdaemon`.  It's not really necessary for daemons at this stage, but it might be useful for profiling long running tests in the future.

ghubstan committed:
Delete commented statement

ghubstan committed:
Do not run dummy test from driver

ghubstan committed:
Fix varible names

ghubstan committed:
Remove test scaffolding logic from ApiTestConfig

ghubstan committed:
Refresh dao-setup files in Scaffold setup

Many test cases will need to run on a clean, initialized dao-setup;
we do not want call dao-setup.gradle tasks from Java.

For now, every JUnit test case will refresh the dao-setup data in
@BeforeClass, but @AfterClass will not clean out the dao-setup files.
We want to be able to look at bitcoin-core and bisq app data & log
files after a test.

Given the difficulty of refreshing dao-setup files using Java,
I took a short cut and used the apitest.linux pkg's BashCommand
to replace dao-setup files.

Some bitcoin.conf and blocknotify configuratation logic was
also moved from ApiTestConfig to Scaffold in this commit.

ghubstan committed:
Remove hacked method tests

Although JUnit is geared towards running randomly ordered unit
tests, we can use it to write API tests to be run in a pre-defined
order.  This is necessary because API tests will be functional
(narrow to broad) and end-to-end;  we want to be able to write
@Test methods that run in @Order.

The next commit will add a JUnit OrderedRunner class, and an @Order
annotation.

ghubstan committed:
Support @Order-ing of JUnit tests

ghubstan committed:
Add first JUnit 'method' tests

These @Order-ed test cases should be run with OrderedRunner.

They are called 'method' tests because they are the closest
thing to an API 'unit' test.  Method tests are not end to end
tests;  they have access to gRPC service stubs and do asserts
on Java values.  We can write 'method' tests for the new
gRPC methods we need to implement (test driven development).

Some notes about the three API test categories...

The plan is to categorize some of the narrower functional tests
as 'scenario' tests.  Like 'method' tests, 'scenario' tests will
have  access to gRPC service stubs and do asserts on Java values,
but they will cover use cases requiring several gRPC and bitcoin-cli
method calls.  One example is funding a wallet:  using bitcoin-cli
and bisq-cli to generate new addresses, send BTC from bitcoin-core
to a Bisq wallet, generate a bitcoin block, and check the Bisq
wallet's balance.  Another example is wallet unlocking/locking
scenarios that are a bit more complex than 'unit' or 'method' tests.

The third category of tests will be end to end, and live in an
'e2e' package.  They will cover more complex use cases and
do asserts on the Bisq CLI's console output (stdout/stderr),
not return values from gRPC service stubs.

There may sometimes be a fine line between what is a 'scenario'
test and an 'e2e' test.  But what starts out as a 'scenario' test
during test driven development can migrate to an 'e2e' test after
thorough testing in the quicker 'scenario' dev/test cycle.

ghubstan committed:
Add driver for running method tests

API test cases are not in a maven/gradle project test folder.
IDEs may not automatically configure JUnit test launchers,
and a gradle build command will not automatically run tests.
This class is provided as a convenience until gradle tasks
for running test cases are implemented.

ghubstan committed:
Get rid of references to removed @Skip annotation

ghubstan committed:
Remove @Skip annotaion

We are using JUnit now and can use @Ignore

ghubstan committed:
Fix codacy problem in bash script

Double quote to prevent globbing and word splitting.

ghubstan committed:
Replace config 'numSetupTasks' with 'supportingApps'

This change replaces the non-intuitive numSetupTasks=<Integer>
configuration option with a supportingApps=<String> option.  It
accepts a comma delimited list of app names, and determines
which background apps will be started in @BeforeClass.

None of the current method tests need all supporting apps, and
this change will reduce scaffolding setup and teardown time.

The current method test cases were changed to use this option.
WalletProtectionTest and GetVersionTest only need to start "alicedaemon".
GetBalanceTest needs "bitcoind,seednode,arbdaemon,alicedaemon".

ghubstan committed:
Allow more time for background app shutdown

ghubstan committed:
Refactor ApiTestCase class hierarchy

Also added a new base ScenarioTest class, which extends
MethodTest;  Method and Scenario tests cases have access to gRPC
service stubs, but end to end test cases should never use them.

ghubstan committed:
Add wallet protect method tests

Some of these 'method' tests make more than one gRPC call, but
they are for checking correctness of a single gRPC method,
and don't quite fall into the 'scenario' test category.

ghubstan committed:
Add tests for resetting a wallet password

ghubstan committed:
Add --bisqAppInitTime=<Long> config option

This change makes configurable the amount of time (ms) each Bisq
instance is given to initialize before starting (a) another Bisq
instance, or (b) tests.

dmos62 commented on apitest/src/main/java/bisq/apitest/Scaffold.java:
```diff
@@ -0,0 +1,445 @@
+                + "regtest=1\n"
+                + "[regtest]\n"
+                + "peerbloomfilters=1\n"
+                + "rpcport=18443\n"
```
rpcport is configured in ApiTestConfig, but here it's hardcoded.

ghubstan replied:
Oops.  Fixed it and two other hardcoded configs in commit cf3b545.

ghubstan committed:
Fix hardcoded bitcoin.conf property values

ghubstan committed:
Fix comment and code styling

ghubstan committed:
Fix 'bitcoind not found' error message

ghubstan committed:
Delete throws clause from method signature

ghubstan committed:
Fix bitcoind startup error handling

Like Bisq instance startup, bitcoind startup should save any error msg
from bash, print the error msg, and System.exit(1).

ghubstan committed:
Bump bisqAppInitTime default back up to 5s

mrosseel commented:
some small questions from my side:
- how hard is it to make it cross-platform? Just converting the .sh scripts or something more?
- how hard is it to integrate this into travis ?

dmos62 commented:
@mrosseel native Windows compatibility would take a fair amount of work and was a non-goal from the start. Running Linux environments under Windows is getting easier every day though, so I'm sure one could find a way around this without significant modifications.

ghubstan commented:
> * how hard is it to make it cross-platform? Just converting the .sh scripts or something more?

1-2 days of work, I estimate -- for someone with a Windows machine.   Right now, there are 22 in-line bash commands executed from Java (only 1 separate `.sh` script ).   To work on Windows, the `bisq.apitest.linux.BashCommand` would need to be ported to `bisq.apitest.windows.WindowsCommand`, and 22 `BashCommands` would need to be ported to `WindowsCommands`.  

 I would be surprised if this did not already work on OSX if bash is on the system. 

> how hard is it to integrate this into travis ?

`[EDITED 19-July-2020]`

Right now, travis would need to run a normal build + a task to download & extract dao-setup.zip files:

    ./gradlew clean build :apitest:installDaoSetup

Then run API `@Test` cases:

    ./gradlew :apitest:test -DrunApiTests=true

ghubstan commented:
> @mrosseel native Windows compatibility would take a fair amount of work and was a non-goal from the start. Running Linux environments under Windows is getting easier every day though, so I'm sure one could find a way around this without significant modifications.

@dmos62 may be more realistic in his estimation of the work effort due to the differences between linux bash and windows system commands.

ghubstan committed:
Remove all sudo related logic from the linux pkg

The BashCommand wrapper should never be used run run sudo commands,
and if the word "sudo" is found in a bash command, System.exit(1)
will be called.

ghubstan committed:
Do not "killall bitcoind" processes

Just kill the one we started.

ghubstan committed:
Create convenient way to call bitcoin-cli from tests

A BitcoinCliHelper class was added.  GetBalanceTest's @BeforeClass
was simplified, and it no longer starts up the arbitration node.

ghubstan committed:
Add FundWalletScenarioTest

Some refactoring was done to reduce some of the boilerplate.

ghubstan committed:
Add JUnitHelper to run tests from JUnitCore

dmos62 commented on apitest/src/main/java/bisq/apitest/scenario/FundWalletScenarioTest.java:
```diff
@@ -0,0 +1,78 @@
+        bitcoinCli.sendToAddress(unusedAddress, btc);
+
+        bitcoinCli.generateBlocks(1);
+        sleep(1500);
```
`bitcoind` has the `blocknotify` callback for this. Not really a problem now, but something to consider if the number of "sleeps" grows cumbersome.

ghubstan replied:
This will become a nuisance, but right now I don't know of another way to give `:daemon` some time to parse new blocks.  (The callback does not reach the `:cli`.)

dmos62 replied:
Oh right, we're waiting for `:daemon`, not `bitcoind`. `:daemon` will probably have a gRPC endpoint for getting notifications about funding events.

dmos62 commented on apitest/src/main/java/bisq/apitest/method/GetBalanceTest.java:
```diff
@@ -0,0 +1,68 @@
+    public void testGetBalance() {
+        var balance = grpcStubs.walletsService.getBalance(GetBalanceRequest.newBuilder().build()).getBalance();
+        assertEquals(1000000000, balance);
+    }
```
The wallet being funded with 10 BTC is an implicit step in the setup, which feels wrong. Preferably it would be explicit. I couldn't track down where it happens. Comment?

ghubstan replied:
It is an important part of the `dao-setup` for regtest/dev mode testing.     Bob & Alice are given initial BTC & BSQ balances and dummy payment accounts so they can trade without coders / testers having to go through the motions of funding wallets from `bitcoin-cli -regest`, setting up payment accounts, etc.  Having to do that would make manual  release testing even more time consuming that it already is.  

There are more details [here](https://github.com/bisq-network/bisq/blob/master/docs/dao-setup.md).

Eventually, we want to create and test the [DAO setup environment](https://github.com/bisq-network/bisq/blob/master/docs/dao-setup.md#setup-a-custom-dao-genesis-transaction), but I don't think that's in scope for quite awhile.  For now, we use the pre-fab [dao-setup.zip](https://github.com/bisq-network/bisq/blob/master/docs/dao-setup.zip) -- what sets up the pre-funded wallets for Bob & Alice.

dmos62 replied:
Do you think it would make automated tests more time consuming as well. Feels dirty not starting with a clean slate. As an alternative, we could be diligent about commenting in the test suites what the initial state is.

ghubstan replied:
I don't know what the cost would be until I could compare.  There would be a trade off because it would skip the dao-setup files refresh in `@BeforeClass`.

I added a javadoc comment explaining why Bob & Alice have non-zero initial BTC & BSQ balances and a dummy payment account before tests start.  (See commit b4d3ea7.)

ghubstan committed:
Do not subtract fee from 'bitcoin-cli sendtoaddress'

Test cases need to be as simple as possible for now.

dmos62 commented on apitest/src/main/java/bisq/apitest/scenario/FundWalletScenarioTest.java:
```diff
@@ -0,0 +1,78 @@
+        // so the new wallet balance will be exactly 12.5 btc.
+        // We should calculate the fee based on the fee rate and tx size
+        // instead of hard coding the fee amount.
+        double btc = parseDouble("2.5") + parseDouble("0.0000336");
```
We could use the `subtractfeefromamount` parameter set to false to not have to calculate a fee.

ghubstan replied:
This is fixed in commit 5d7133a.

dmos62 commented on apitest/src/main/resources/bitcoin.conf:
```diff
@@ -0,0 +1,16 @@
+rpcpassword=apitest
+
+# The blocknotify path will be defined and at runtime.
+# blocknotify=bash ~/.bitcoin/blocknotify %
```
It might be worth it to not use a bitcoin.conf at all and pass all parameters inline, like this: https://gist.github.com/dmos62/20981dae799e2fc31b62dd443c44c7ae#file-tests-bats-L139

Since your setup already works, this is low priority.

ghubstan replied:
Passing all `bitcoind` parameters inline is cleaner.  It's one less file the test harness has to write & install, and the `bitcoind` configuration is clearly visible in the 'apitest' logfile.   Fixed in commit f7d8c0e.

ghubstan committed:
Remove final modifier

ghubstan committed:
Do not use bitcoin.conf files

This change removes code for installing a regtest bitcoin.conf file.
It also removes an unused  bitcoin.conf file from the main resources
directory.

Now, the bitcoind startup command passes all configurations on the
command line.  (See bisq.apitest.linux.BitcoinDaemon.java)

ghubstan committed:
Delete unused bitcoin.conf from resources dir

ghubstan committed:
Delete all JUnit related class from main sources

JUnit tests are being moved to the subproject's test sources
directory for the sake of convention -- for developers using
this test harness, and for IDE and gradle JUnit integration.

dmos62 commented on apitest/src/main/resources/blocknotify:
```diff
@@ -0,0 +1,20 @@
+echo $1 | nc -w 1 127.0.0.1 5123
+
+# Some other node listens on port 5124, etc.
+# echo $1 | nc -w 1 127.0.0.1 5124
```
Not requesting this change, but note that you probably can pass a blocknotify script to bitcoind inline, like this: https://gist.github.com/dmos62/20981dae799e2fc31b62dd443c44c7ae#file-tests-bats-L150 (untested).

ghubstan replied:
I had thought of that;  we could pass an inline script for notifying only the ports being used by the scaffolding to avoid errors in the bitcoind log.  It's on the TODO list.

ghubstan committed:
Moving GrpcStubs to test sources

ghubstan committed:
Move test cases into subproject test sources

This change reorganizes the ':apitest' subproject to conform to a
typical gradle project, where all JUnit test cases are located in
the subproject's test sources folder.  This makes running tests
from an IDE or gradle command line interface work as expected.
It will also help keep Travis CI configuration simple.

To avoid interfering in normal builds, the gradle ':apitest test' task
is disable by default;  API tests will only run when a '-Dforce-true'
system property is passed to gradle.

To run API tests, run a normal build and install dao-setup files:

    ./gradlew clean build :apitest:installDaoSetup

Then run the tests:

    ./gradlew :apitest:test -Dforce=true

Try to avoid adding the '-Dforce=true' property to any other gradle
tasks, because this enables the ':apitest test' task, and would kick
off API tests before a normal build completed.

The build.gradle file was modified to support this code reorg, and
the 'org.junit.jupiter' dependendency was upgraded to v5.6.2 -- only
in the ':apitest:test' dependency definiitions, not anywhere else in
the bisq dependency definitions.  The upgrade is necessary for
running ordered tests.

Since the scaffolding may be set up from either test cases (under the
test src folder), or a class under the main src folder, some changes
were made to ensure configuration paths are correct for either use
case.  For example, when the 'bisq-apitest' script is run from the root
project directory, the current working directory is the root project
directory.  When gradle or an IDE is used to run @Test cases, the
current working directory is :apitest subproject directory.

The main source's ApiTestMain class has been stripped down, and exists
only to keep the gradle build happy -- it needs a 'mainClassName'
property.  But this main driver does have uses.  See the class comments.

The other changes in this commit were made to fix style and syntax
problems.

ghubstan committed:
Change :apitest:test task system property name

The property name 'force' is changed 'runApiTests'.

To run the :apitest test task:

    ./gradlew :apitest:test -DrunApiTests=true

ghubstan committed:
Add comment about Bisq DAO dev environment

Some explanation is needed about why Bob & Alice have non-zero
initial BTC and BSQ balances when tests run.  The comments
also include links to more detailed information about the DAO/
regtest testing environment.

ghubstan committed:
Fix @BeforeClass error handling and use jupiter api

The Scaffold set up was calling System.exit(1) when it encountered a
configuration error, or a problem starting bitcoind & bisq instances.
This incorrect error handling was hiding errors from gradle, and
preventing tests that would otherwise successfully complete.

This change fixes the problem by throwing an IllegalStateException up
to the test case's @BeforeClass method -- to be caught and used in a
JUnit fail(ex) call.  An 'initializationError' triggered in @BeforeClass
correctly bubbles up to gradle, and will not block execution of remaining
tests.  A gradle Test Summary containing any initialization errors is also
produced in <project-dir>/apitest/build/reports/tests/test/index.html

This change also fixes many import statements and asserts to ensure
'org.junit.jupiter.api.*' is used in place of 'org.junit.*', for
proper gradle integration.

ghubstan committed:
Remove unnecessary curly braces

ghubstan committed:
Change 'missing bitcoind path' error msg

Since JUnit tests cannot take program arguments, change the
wording to instruct the user to configure the bitcoind path
in apitest.properties

dmos62 reviewed (dismissed):
ACK.

We had a lot of back and forth over this PR. At 3k lines the PR isn't small. Part of the complexity is the shell-level interaction between bitcoind, bitcoin-cli and Bisq daemons. I had some reservations about doing everything in Java, since that required implementing shell-type functionality. I was floating an idea to use Java routines for things that require Java API access and Bash routines for the rest (like setup, teardown and e2e tests), hoping that that would cut down on the boilerplate (and would be possible at all). I wrote up a [Bash e2e testing harness](https://gist.github.com/dmos62/20981dae799e2fc31b62dd443c44c7ae) for comparison, but it was moot point, because ghubstan's PR was almost done by then. Some of the ideas there were implemented in this PR. It seems to work pretty well right now. Some kinks may (or may not) appear when the number of tests gets high, but we'll cross that bridge when we get there.

Thank you @ghubstan for the work!

ripcurlx commented:
@sqrrm Could you please have a look at this PR?

ghubstan committed:
Do not leave orphaned processes after failed teardown

The test harness should not fail a test case's @AfterAll (teardown)
method on the first background instance shutdown exception.  This
change makes the shutdown logic similar to the startup's:  it caches
any exceptions that may have occurred during an instance shutdown,
logs them, then proceeds to shut down the next background instance.
An IllegalStateException (the 1st one) is passed up to @AfterAll method
only after the scaffolding teardown process is complete, to avoid leaving
any orphaned java or bitcoind processes running after a java system
exit.

ghubstan committed:
Remove extra whiteline

ghubstan commented:
@dmos62 , sorry for the late check-in (dismissing your ACK), but it was important.   See commit 27ee4b8.

ghubstan committed:
Clarify scaffold tear down error handling

The Scaffold#tearDown() method was split into two methods.  The
original tearDown() now passes the background process/task array
to a new shutDownAll() method.  This new method loops through the
tasks in a more readable way, plainly expressing the intent to log
all shutdown exceptions for each process being shut down, but not
throwing an exception while processes are being shut down.
The new shutDownAll() method returns the first shutdown exception
encountered, which in turn is passed up to the test case's @AfterAll
method.

ghubstan commented:
> semantic duplication

Fixed in commit 8bb7e12

dmos62 reviewed (dismissed):
ACK

ghubstan committed:
Add fallbackfee param to bitcoind start cmd

This commit adds a -fallbackfee=0.0002 parameter to the start 'bitcoind'
command to keep bitcoin-core v0.20.1 working as v0.19.1 does -- with the
'fallbackfee' enabled.

Bitcoin v0.20.0 contains a fix for inconsistent behaviour related
to this fallbackfee configuration.  Prior to v0.20, fallbackfee
was disabled (0) by default for the mainnet chain, but enabled
(0.0002) for the testnet and regtest chains.

A test case with bitcoin-core v0.20.1 was breaking on
the bitcoin-cli 'sendtoaddress' command, which was returning an
error message instead of a tx-id:

    error code: -4
    error message:
    Fee estimation failed. Fallbackfee is disabled. \
    Wait a few blocks or enable -fallbackfee.

Bitcoin-core v0.20.0 release notes contain info about this change:
    https://bitcoin.org/en/release/v0.20.0#updated-rpcs-1

The Bitcoin-core PR is https://github.com/bitcoin/bitcoin/pull/16524

dmos62 commented on apitest/src/main/java/bisq/apitest/linux/BitcoinDaemon.java:
```diff
@@ -55,6 +55,7 @@ public void start() throws InterruptedException, IOException {
                 + " -txindex=1"
                 + " -peerbloomfilters=1"
                 + " -debug=net"
+                + " -fallbackfee=0.0002"
```
This change seems to change the behaviour of bitcoind in mainnet mode. Before 0.20.0 bitcoind had different default `fallbackfee`s for mainnet and test modes (0 and 0.0002 respectively), which was then fixed by setting the defaults to 0 in all cases.

By setting `fallbackfee=0.0002` we're setting it on all net modes. I can't tell if that's a problem.

dmos62 replied:
Oh wait. This bitcoind will never run in mainnet, will it? Then the concern is that the apitest testing harness will have a different bitcoind environment than a production daemon would.

Again, I don't fully understand the significance of the `fallbackfee` option. It's used when the bitcoind can't estimate a "floating fee", right? And it only applies when sending funds from a wallet managed by the bitcoind instance, right? If both answers are yes, then I don't see this causing differences between prod and dev environments, but I'd like someone else to sanity check this too.

ghubstan replied:
This new fallbackfee setting would change the behaviour of bitcoind in mainnet mode -- from the default 0.00 to 0.0002.  But `apitest` cannot run on mainnet unless the code is hacked, and `apitest` should never run on mainnet -- it should never spend mainnet BTC, do a trade on mainnet, etc.

>> Then the concern is that the apitest testing harness will have a different bitcoind environment than a production daemon would.

Yes, it does mean that the `apitest` harness will have a different `bitcoind` environment than a production daemon would, but that is what we want.  The bitcoin-core change ([PR 16524](https://github.com/bitcoin/bitcoin/pull/16524)) was to make the fallbackfee the same on all chains -- to fix bitcoin-core's inconsistency.  It was 0 (disabled) on mainnet, but 0.0002 for testnet/regtest chains.  Now (v0.20+), it is the same for all chains: 0 (disabled) by default for all chains.

We want bitcoin-core v0.20 (regtest) to behave like it did before the bitcoin-core PR 16524 change.

In the context of `apitest`, the fallbackfee affects the behaviour of the bitcoin-cli `sendtoaddress` command, where we use the `subtractfeefromamount=false` parameter.  (See `FundWalletScenarioTest#testFundWallet()`)  This `bitcoin-cli` command errors if the fallbackfee is disabled on bitcoin-core v0.20.  Enabling it on `bitcoind` ensures this command behaves the same way it does on v0.19.

sqrrm replied:
I think this is reasonable. These tests should never be used to start a live bitcoind.

ghubstan committed:
Fix BitcoinCli wrapper error handling

This change checks the system call exit status of bitcoin-cli
commands, and populates a new error message accessor if the
system exist status != 0.

dmos62 commented on apitest/src/test/java/bisq/apitest/method/BitcoinCliHelper.java:
```diff
@@ -40,9 +39,12 @@ public BitcoinCliHelper(ApiTestConfig config) {
+            BitcoinCli newAddress = new BitcoinCli(config, "getnewaddress").run();
+
+            if (newAddress.isError())
+                fail(format("Could generate new bitcoin address:%n%s", newAddress.getErrorMessage()));
```
Typo? "Could _not_ generate ..."?

ghubstan replied:
Oops.  Fixed in commit 9637cc0.

ghubstan committed:
Fix test fail() msg

sqrrm commented:
Finally managed to get the tests to pass. There is a conflict with current running bitcoin core processes it seems. Once my local bitcoin-qt regtest instance was shut down the tests passed.

ghubstan committed:
Use non-default regtest bitcoind -rpcport

The default bitcoind / bitcoin-cli rpcport option has been changed
from 18443 to 19443, to help avoid rpcport conflicts between apitest's
bitcoind instances and other bitcoind and/or bitcion-qt instances
which are probably using the bitcoin-core default (regtest) rpcport
18443.

However, this commit cannot include other changes for avoiding bind
address:port conflicts between apitest bitcoind instances and other
regtest bitcoin-core instances because bitcoinj's bind port is hardcoded
in RegTestParams.java as 18444.

In order to avoid bitcoin-core regtest mode bind address conflicts,
you must start or restart your bitcoind or bitcoin-qt instance with a
non-default bind port argument, e.g.

	bitcoin-qt -regtest -port=20444

ghubstan committed:
Add regtest-port-conflicts.md doc

ghubstan commented:
> There is a conflict with current running bitcoin core processes it seems. Once my local bitcoin-qt regtest instance was shut down the tests passed.

Commit 72ff4dc resolves the `-rpcport` conflict, but the bind address:port conflict is still unresolved because `bitcoinj` hardcodes the regtest bind address:`port` in `RegTestParams`.  Commit e88bdb9 adds a doc explaining how to work around the conflict by starting your bitcoin-qt app with a `-port` option:  `$ bitcoin-qt -regtest -port=20444`

ghubstan committed:
Fix port number typo

ghubstan committed:
Add build / run / test categories docs

ghubstan committed:
Explain how to run test cases from Intellij

ghubstan committed:
Fix typo

ghubstan committed:
Update README

sqrrm commented on apitest/src/test/java/bisq/apitest/method/MethodTest.java:
```diff
@@ -0,0 +1,89 @@
+                .findFirst()
+                .get()
+                .getAddress();
+
```
```suggestion
```

ghubstan replied:
Extra while line removed in commit c3abd4e.

sqrrm commented on apitest/src/test/java/bisq/apitest/method/WalletProtectionTest.java:
```diff
@@ -0,0 +1,137 @@
+        Throwable exception = assertThrows(StatusRuntimeException.class, () ->
+                grpcStubs.walletsService.lockWallet(request));
+        assertEquals("UNKNOWN: wallet is already locked", exception.getMessage());
+
```
```suggestion
```

ghubstan replied:
Extra while line removed in commit c3abd4e.

sqrrm reviewed (dismissed):
ACK

sqrrm commented:
I had some suggested empty line removals, not critical. The codacy complaints don't seem legit. We have been requiring non-empty blocks, but I think that's not a good policy so hope to remove that together with the fix of bad indentation rules.

ghubstan committed:
Remove white lines

ghubstan commented:
Closing/reopening PR to force travis build after [pricenode unit test failure](https://travis-ci.org/github/bisq-network/bisq/builds/718958332) .

ghubstan commented:
Closing/reopening PR again to force codacy check after @ripcurlx made some codacy config changes.

ghubstan commented:
> Closing/reopening PR again to force codacy check...

Closing/reopening PR does not force codacy check.

ghubstan committed:
Add punctuation & re-phrase sentence in README

A new commit was needed to force a codacy check after changes were
made to codacy rules.

ghubstan committed:
Change access modifer

This commit is for forcing a codacy check.  The previous
change to an .md doc did not force a codacy check.

ghubstan reviewed (commented):
RE 2ba0ee9:  the private method was also moved to the bottom of the class.

This change did not force a codacy check.

sqrrm reviewed (dismissed):
ACK

ghubstan committed:
Put 'empty' comments inside ignored catch blocks

Follow codacy rule against empty blocks.

ghubstan commented:
This PR should be merged before [PR 4417](https://github.com/bisq-network/bisq/pull/4417).

ripcurlx commented:
@ghubstan I contacted Codacy on this issue. As the configuration problem should be solved (I've even deactivated it for testing), but it sill doesn't green light this PR.

ghubstan committed:
Shorten line length < 120 chars

ghubstan commented:
> @ripcurlx (& @sqrrm):  ghubstan I contacted Codacy on this issue. As the configuration problem should be solved (I've even deactivated it for testing), but it sill doesn't green light this PR.

I pushed a change to shorten some too-long lines in one class, and codacy passed it.   But I don't know if the problems are fixed yet, since codacy  analyzed only the one changed class(?)  I may push a small change to the next [PR](https://github.com/bisq-network/bisq/pull/4417) , based off this one, to see what codacy does with it.

By the way, the failed Travis CI JDK12 test failed to install JDK 12, so I won't close/re-open this PR to force another  Travis build.    I don't want to risk losing the Codacy Review PASS ;)

sqrrm reviewed (approved):
ACK
